CircuitPython_Pyloton/pyloton.py

```diff
@@ -0,0 +1,563 @@
+                advs[adv.address] = adv
+
+        self.ble.stop_scan()
+        self._status_update("Speed and Cadence: Stopped scanning")
+        if not advs:
+            # Nothing found. Go back and keep looking.
+            return []
+
+        # Connect to all available CSC sensors.
+        self.cyc_connections = []
+        for adv in advs.values():
+            self.cyc_connections.append(self.ble.connect(adv))
+            self._status_update("Speed and Cadence: Connected {}".format(len(self.cyc_connections)))
+
+
+        self.cyc_services = []
+        for conn in self.cyc_connections:
+
+            self.cyc_services.append(conn[CyclingSpeedAndCadenceService])
+        self._status_update("Pyloton: Finishing up...")
+        return self.cyc_connections
+
+
+    def _speed_helper(self, values):
+        wheel_diff = values.last_wheel_event_time - self._previous_wheel
+        rev_diff = values.cumulative_wheel_revolutions - self._previous_revolutions
+
+        if wheel_diff:
+            # Rotations per minute is 60 times the amount of revolutions since
+            # the last update over the time since the last update
+            rpm = 60*(rev_diff/(wheel_diff/1024))
+            # We then mutiply it by the wheel's circumference and convert it to mph
+            speed = round((rpm * self.circumference) * (60/63360), 1)
+            if speed < 0:
+                speed = self._previous_speed
+            self._previous_speed = speed
+            self._previous_revolutions = values.cumulative_wheel_revolutions
+            self._i = 0
+        else:
+            self._i += 1
+            if self._i >= 3:
+                speed = 0
+        self._previous_wheel = values.last_wheel_event_time
+
+
+    def _cad_helper(self, values):
+        crank_diff = values.last_crank_event_time - self._previous_crank
+        crank_rev_diff = values.cumulative_crank_revolutions-self._previous_rev
+
+        if crank_rev_diff:
+            # Rotations per minute is 60 times the amount of revolutions since the
+            # last update over the time since the last update
+            cadence = round(60*(crank_rev_diff/(crank_diff/1024)), 1)
+            if cadence < 0:
+                cadence = self._previous_cadence
+            self._previous_cadence = cadence
+            self._previous_rev = values.cumulative_crank_revolutions
+            self._j = 0
+        else:
+            self._j += 1
+            if self._j >= 3:
+                cadence = 0
+        self._previous_crank = values.last_crank_event_time
+
+
+    def read_s_and_c(self):
+        """
+        Reads data from the speed and cadence sensor
+        """
+        speed = self._previous_speed
+        cadence = self._previous_cadence
+        for conn, svc in zip(self.cyc_connections, self.cyc_services):
+            if not conn.connected:
+                speed = cadence = 0
+                continue
```

Comment:
This will loop forever when there are no connections. Maybe return instead so caller can take note of connection loss?

Reply:
It looks like am wrong about this. I thought it was a `while True` for some reason.

Reply:
But if only one of two sensors is not connected, we'll lose data for both? Is that OK?

Reply:
Personally, I think that's alright. This is just put in there because a sensor may disconnect after the main loop in code.py checks if it's still connected, and that can cause read_s_and_c to fail and give an ugly error message. This just sets both to zero and then the main loop should be broken out of on the next run.

Reply:
I understand now. Resolving this convo.